lib/div.cpp

```diff
@@ -0,0 +1,457 @@
+
+  at::ScalarType out_dtype = at::promote_types(a.scalar_type(), b.scalar_type());
+  at::Tensor out = at::empty_like(a, at::TensorOptions().dtype(out_dtype).device(a.device()));
+
+  std::string kernel_name = use_integer_kernel ? INT_KERNEL_NAME : FLOAT_KERNEL_NAME;
+  const TritonJITFunction& f =
+      TritonJITFunction::getInstance((utils::get_triton_src_path() / "div.py").string(), kernel_name);
+
+  int64_t tile_size = 1024;
+  int64_t n = out.numel();
+  int num_blocks = (n + tile_size - 1) / tile_size;
+
+  c10::cuda::CUDAStream stream = c10::cuda::getCurrentCUDAStream();
+  CUstream raw_stream = static_cast<CUstream>(stream.stream());
+  c10::DeviceGuard guard(out.device());
+  // launch Triton kernel
+  f(raw_stream, num_blocks, 1, 4, 2, 0, a, b, out, n, tile_size);
+
+  return out;
+}
+
+at::Tensor floor_div(const at::Tensor& a_, const at::Tensor& b_) {
+  // ========== Scalar // Scalar ==========
+  if (a_.dim() == 0 && b_.dim() == 0) {
+    if (a_.is_floating_point() || b_.is_floating_point()) {
+      double r = std::floor(a_.item<double>() / b_.item<double>());
+      return torch::tensor(r, a_.options());
+    } else {
+      int64_t a_val = a_.item<int64_t>();
+      int64_t b_val = b_.item<int64_t>();
+      int64_t q = a_val / b_val;
+      int64_t r = a_val % b_val;
+      int64_t fix = (r != 0 && ((a_val < 0) != (b_val < 0))) ? 1 : 0;
+      return torch::tensor(q - fix, a_.options());
+    }
+  }
+  // ========== Tensor // Scalar ==========
+  if (a_.dim() > 0 && b_.dim() == 0) {
+    at::Tensor b_expand = at::full_like(a_, b_.item<double>());
+    bool is_int = c10::isIntegralType(a_.scalar_type(), true) && c10::isIntegralType(b_.scalar_type(), true);
+    return launch_floor_div_kernel(a_, b_expand, is_int);
+  }
+  // ========== Scalar // Tensor ==========
+  if (a_.dim() == 0 && b_.dim() > 0) {
+    at::Tensor a_expand = at::full_like(b_, a_.item<double>());
+    bool is_int = c10::isIntegralType(a_.scalar_type(), true) && c10::isIntegralType(b_.scalar_type(), true);
+    return launch_floor_div_kernel(a_expand, b_, is_int);
+  }
+  // ========== Tensor // Tensor ==========
+  bool is_int = c10::isIntegralType(a_.scalar_type(), true) && c10::isIntegralType(b_.scalar_type(), true);
+  return launch_floor_div_kernel(a_, b_, is_int);
+}
+
+at::Tensor launch_floor_div_kernel_(at::Tensor& a_, const at::Tensor& b_, bool use_integer_kernel) {
+  auto res = torch::broadcast_tensors({a_, b_});
+  at::Tensor a = res[0].contiguous();
+  at::Tensor b = res[1].contiguous();
+
+  TORCH_CHECK(a.sizes() == b.sizes(), "Broadcasting failed.");
+  TORCH_CHECK(a.device().is_cuda(), "Only CUDA supported for floor_div");
+
+  at::ScalarType out_dtype = at::promote_types(a.scalar_type(), b.scalar_type());
+
+  if (a_.dtype() != out_dtype) {
+    a_ = a_.to(out_dtype);
+  }
+  if (!a_.sizes().equals(a.sizes())) {
+    a_ = at::empty(a.sizes(), a_.options().dtype(out_dtype));
+  }
```

Comment:
This is incorrect. In the case of inplace operation, the first operand is also the output and it should match sizes of the broadcast shape. If they don't match then fail.